docs/src/developer/design/user_api.rst

```diff
@@ -36,133 +55,359 @@ functional-style and cmake-based
+   cmake_find_dependency(
+      <name>
+      VERSION <the_version_you_want>
+      FIND_TARGET <target_representing_package>
+      CMAKE_ARGS <options_it_should_have_been_configured_with>
+   )
+
+In practice, following from the :ref:`ua_package_manager` consideration, these
+functions are envisioned as wrappers over a :term:`package manager`. The main
+goal of the user API is to collect the information needed for the package
+manager to build the dependency and for CMaize to use the dependency the package
+manager builds.
+
+Define Build Targets
+====================
+
+Full discussion: :ref:`designing_cmaizes_add_target_functions`.
+
+Once we have found or built all of the :term:`project's <project>` dependencies
+we can move on to building the :term:`build targets <build target>`. Generally
+speaking, the information needed to build a target depends on the coding
+language of the target. For the purposes of this high-level discussion we focus
+on C++; build targets for most other coding languages will have similar needs.
+For a typical C++ target we need to specify the:
+
+- name of the build target,
+- source files defining the build target's implementation,
+- header files defining the build target's public :term:`API`, and
+- build target's dependencies (including other build targets).
+
+The proposed CMaize APIs are:
+
+.. code-block:: CMake
+
+   # Declaring a build target for a (C++) library
+   cmaize_add_library(
+      <name>
+      SOURCE_DIR <where_the_source_files_are_located>
+      INCLUDE_DIRS <directories_containing_header_files>
+      DEPENDS <dependency0> <dependency1> ...
+   )
+
+   # Declaring a build target for a (C++) executable is similar
+   cmaize_add_executable(
+      <name>
+      SOURCE_DIR <where_the_source_files_are_located>
+      INCLUDE_DIRS <directories_containing_header_files>
+      DEPENDS <dependency0> <dependency1>
+   )
+
+Like the "Find Dependencies" step before it, the APIs for defining build targets
+are designed primarily for collecting information pertaining to the build
+target. Unlike the "Find Dependencies" step the backend of API calls for
+defining build targets is CMake. The result of calling these methods are
+properly configured CMake targets.
+
+Test Project
+============
+
+After targets are built the next step is to test that they were built correctly.
+Testing build targets with CMake often requires:
+
+- finding dependencies of the testing framework,
+- building testing targets which consume the project's build targets, and
+- registering the tests with CTest.
+
+It is also worth noting that tests are often built conditionally (e.g., a
+:term:`build system` typically does not build the tests of dependencies built
+during the "Find Dependencies" step). To that end CMake defines the
+``BUILD_TESTING`` variable; when set to ``TRUE`` tests are built, otherwise they
+are not. Proposed CMaize APIs for testing a project:
+
+.. code-block:: CMake
```

Comment:
Are these mostly proposed interfaces that do not exist right now? I believe the only one we have is `cmaize_add_tests`, and CTest's `add_test` command is not necessary to register the test. That is done internally in `cmaize_add_tests`.